swri_image_util/src/nodelets/replace_colors_nodelet.cpp

```diff
@@ -0,0 +1,457 @@
+    }
+    else if (!colormap_specified)
+    {
+      ROS_ERROR("Color transformation was not specified. Images will ");
+      ROS_ERROR("only be converted to their gray scale equivalents");
+    }
+
+    // Set up the ROS interface
+    image_transport::ImageTransport it(nh);
+    image_pub_ = it.advertise("modified_image", 1);
+    image_sub_ = it.subscribe(
+      "image",
+      1,
+      &ReplaceColorsNodelet::imageCallback,
+      this);
+  }
+
+  // Callback for getting the input image to change the colors on
+  void ReplaceColorsNodelet::imageCallback(
+    const sensor_msgs::ImageConstPtr& image_msg)
+  {
+    // Only do the color conversion if someone is subscribing to the data
+    if (image_pub_.getNumSubscribers() == 0)
+    {
+      return;
+    }
+
+    // This node currently only support changing gray scale images
+    if (image_msg->encoding != sensor_msgs::image_encodings::MONO8)
+    {
+      ROS_ERROR("Changing image colors is only supported for MONO8 images");
+      return;
+    }
+
+    // Convert image data from ROS to OpenCV type
+    cv_bridge::CvImageConstPtr original_image = cv_bridge::toCvShare(image_msg);
+
+    // Allocate space for the modified image
+    cv::Mat modified_image = cv::Mat::zeros(
+      original_image->image.rows,
+      original_image->image.cols,
+      CV_8UC3);
+
+    // Do the actual color replacement
+    swri_image_util::replaceColors(
+      original_image->image,
+      color_lut_,
+      modified_image);
+
+    // Copy results to output message and set up the header and encoding values
+    cv_bridge::CvImagePtr output = boost::make_shared<cv_bridge::CvImage>();
+    output->image = modified_image;
+    output->encoding = sensor_msgs::image_encodings::RGB8;
+    output->header = image_msg->header;
+
+    // Publish the modified image to the rest of the system
+    image_pub_.publish(output->toImageMsg());
+  }
+
+  // Initialize grayscale lookup table
+  void ReplaceColorsNodelet::initLut()
+  {
+    // Sets every row in the lookup table to a triple <x, x, x>,
+    // where x is the grayscale value. This will directly map
+    // the gray values to their equivalent RGB representation
+    for (uint32_t idx; idx < NUM_GRAY_VALUES; idx++)
+    {
+      color_lut_.at<cv::Vec3b>(0, idx) = cv::Vec3b(idx, idx, idx);
+    }
+  }
+
+  // Read in the colormap and parameters the parameter server has for this
+  // node
+  void ReplaceColorsNodelet::readColormap(const XmlRpc::XmlRpcValue& param)
+  {
+    // This is a multistep process, and if any step goes wrong then the rest
+    // of the process should be aborted.
+    bool success = true;
+
+    // The colormap parameters should be formatted like:
+    // ["colormap_name", num_colors]
+    // This checks to make sure the parametre is a two element array
+    if ((param.getType() != XmlRpc::XmlRpcValue::TypeArray) ||
+      (param.size() != 2))
+    {
+      ROS_ERROR("Colormap specification must be a two entry array");
```

Comment:
What would make it right? Tell the user; e.g.,

    ROS_ERROR("  [\"colormap_name\", num_colors]");